### PR TITLE
[RISCV] Implement tail call optimization in machine outliner

### PR DESCRIPTION
**Pull request by mag-mga**
Following up issue #89822, this patch adds opportunity to use tail call in machine outliner pass.
Also it enables outline patterns with X5(T0) register.

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/machine-outliner-position.mir [hunk 1]
```diff
@@ -1,3 +1,4 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 2
 # NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
```
Comment (wangpc-pp) [inline llvm/test/CodeGen/RISCV/machine-outliner-position.mir:2]: Why is this line duplicated?

Comment (mga-sc) [reply llvm/test/CodeGen/RISCV/machine-outliner-position.mir]: Just misspell. Fixed.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 2]
```diff
@@ -2937,19 +2938,47 @@ bool RISCVInstrInfo::shouldOutlineFromFunctionByDefault(
   return MF.getFunction().hasMinSize();
 }
 
+static bool IsCandidatePatchable(const MachineInstr &MI) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2941]: Function names should start with a lower case letter

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 3]
```diff
@@ -2937,19 +2938,47 @@ bool RISCVInstrInfo::shouldOutlineFromFunctionByDefault(
...
+         F.hasFnAttribute("patchable-function-entry");
+}
+
+static bool CannotInsertTailCall(const MachineInstr &MI) {
+  if (MI.isTerminator())
+    return IsCandidatePatchable(MI);
+  return true;
+}
+
+static bool MIUseX5(const MachineInstr &MI, const TargetRegisterInfo *TRI) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2955]: Use -> Uses

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 4]
```diff
@@ -2937,19 +2938,47 @@ bool RISCVInstrInfo::shouldOutlineFromFunctionByDefault(
...
 std::optional<std::unique_ptr<outliner::OutlinedFunction>>
 RISCVInstrInfo::getOutliningCandidateInfo(
     const MachineModuleInfo &MMI,
     std::vector<outliner::Candidate> &RepeatedSequenceLocs,
     unsigned MinRepeats) const {
 
-  // First we need to filter out candidates where the X5 register (IE t0) can't
-  // be used to setup the function call.
-  auto CannotInsertCall = [](outliner::Candidate &C) {
+  auto CandidateUseX5 = [](outliner::Candidate &C) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2966]: Use -> Uses

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 5]
```diff
@@ -2961,6 +2990,17 @@ RISCVInstrInfo::getOutliningCandidateInfo(
   for (auto &MI : RepeatedSequenceLocs[0])
     SequenceSize += getInstSizeInBytes(MI);
 
+  if (!CannotInsertTailCall(RepeatedSequenceLocs[0].back())) {
+    // tail function = 8 bytes. Can't be compressed
+    for (auto &C : RepeatedSequenceLocs)
+      C.setCallInfo(MachineOutlinerTailCall, 8);
+
+    // Using tail call we move ret instrunction from caller to calee.
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2998]: instruction*
calle*

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 6]
```diff
@@ -3041,19 +3073,29 @@ void RISCVInstrInfo::buildOutlinedFrame(
...
   MBB.addLiveIn(RISCV::X5);
 
   // Add in a return instruction to the end of the outlined frame.
   MBB.insert(MBB.end(), BuildMI(MF, DebugLoc(), get(RISCV::JALR))
-      .addReg(RISCV::X0, RegState::Define)
-      .addReg(RISCV::X5)
-      .addImm(0));
+                            .addReg(RISCV::X0, RegState::Define)
+                            .addReg(RISCV::X5)
+                            .addImm(0));
```
Comment (ilovepi) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:3085]: is this intentional?

**Review by ilovepi — APPROVED** [review]
Other than @topperc's comments this seems good, modulo a few small nits w.r.t. formatting. Thanks for working on this! 😄

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp (hunk 2 above)
Comment (mga-sc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: done

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp (hunk 3 above)
Comment (mga-sc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: done

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp (hunk 4 above)
Comment (mga-sc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: done

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp (hunk 5 above)
Comment (mga-sc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: done

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp (hunk 6 above)
Comment (mga-sc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: No, get changes back. 
It was changed by `git clang-format` ...

**Review by wangpc-pp — COMMENTED** [review]
LGTM in general, but I'd like to know if we have passed all tests in llvm-test-suite with `-Oz`.

**Comment by mga-sc** [comment]
> LGTM in general, but I'd like to know if we have passed all tests in llvm-test-suite with `-Oz`.

@wangpc-pp , llvm-test-suite with `-Oz` locally has been passed. I don't know how to run it using this git infrastructure

**Comment by wangpc-pp** [comment]
> > LGTM in general, but I'd like to know if we have passed all tests in llvm-test-suite with `-Oz`.
> 
> @wangpc-pp , llvm-test-suite with `-Oz` locally has been passed. I don't know how to run it using this git infrastructure

It's OK as long as you have run llvm-test-suite on qemu and all tests are passed. The LLVM Github CI is not so complete for RISC-V.

**Review by wangpc-pp — APPROVED** [review]
LGTM.

**Comment by mga-sc** [comment]
@topperc , ping. Could we merge this MR? I don't have write access.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 7]
```diff
@@ -2961,6 +2990,17 @@ RISCVInstrInfo::getOutliningCandidateInfo(
   for (auto &MI : RepeatedSequenceLocs[0])
     SequenceSize += getInstSizeInBytes(MI);
 
+  if (!cannotInsertTailCall(RepeatedSequenceLocs[0].back())) {
+    // tail function = 8 bytes. Can't be compressed
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2994]: Why can't it be compressed?

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 8]
```diff
@@ -2961,6 +2990,17 @@ RISCVInstrInfo::getOutliningCandidateInfo(
...
+    for (auto &C : RepeatedSequenceLocs)
+      C.setCallInfo(MachineOutlinerTailCall, 8);
+
+    // Using tail call we move ret instruction from caller to calle.
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2998]: callee*

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 9]
```diff
@@ -3054,6 +3089,13 @@ MachineBasicBlock::iterator RISCVInstrInfo::insertOutlinedCall(
     Module &M, MachineBasicBlock &MBB, MachineBasicBlock::iterator &It,
     MachineFunction &MF, outliner::Candidate &C) const {
 
+  if (C.CallConstructionID == MachineOutlinerTailCall) {
+    It = MBB.insert(It, BuildMI(MF, DebugLoc(), get(RISCV::PseudoTAIL))
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:3093]: PseudoTAIL may overwrite either R6 or R7 if the linker doesn't relax the AUIPC+JALR sequence to JAL. How do we know it is ok to overwrite R6 or R7?

Comment (mga-sc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:3128]: Thanks for the comment, I did not consider the case of the influence of registers x6 and x7 before relaxation. Apparently, even llvm-test-suite didn't get such case. 
I rewrote verification algorithm for tail call. If we modify X6/X7 (Specific register get from enabled extensions info) earlier than reads from X6/X7, than my optimization can be applied.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp (hunk 8 above)
Comment (mga-sc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: Fixed

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp (hunk 7 above)
Comment (mga-sc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: Rewrote it in a more precise form
```cpp
    // tail call = auipc + jalr in the worst case without linker relaxation.
    CallOverhead = 4 + InstrSizeCExt;
```

**Comment by mga-sc** [comment]
@topperc , ping

**Review by topperc — APPROVED** [review]
LGTM

**Comment by mga-sc** [comment]
@topperc , could you please merge this PR?

**Pull request merged**